modelforge/potential/dimenet.py

```diff
@@ -0,0 +1,359 @@
+
+        # Gather atomic numbers for neighbor pairs
+        # Z_i and Z_j have shape (nr_of_pairs)
+        Z_i = inputs.atomic_numbers[pairlist_output.pair_indices[0]]
+        Z_j = inputs.atomic_numbers[pairlist_output.pair_indices[1]]
+
+        # Get embeddings for atomic numbers
+        # x_i and x_j have shape (E, emb_size)
+        x_i = self.embeddings(Z_i)
+        x_j = self.embeddings(Z_j)
+
+        # Concatenate embeddings and transformed rbf
+        # x has shape (E, 3 * emb_size)
+        x = torch.cat([x_i, x_j, rbf], dim=-1)
+
+        # Final transformation
+        # x: (E, 3 * emb_size) -> (E, emb_size)
+        x = self.dense(x)
+
+        return x
+
+
+class Envelope(nn.Module):
+    """
+    Envelope function that ensures a smooth cutoff.
+    """
+
+    def __init__(self, exponent: int):
+        super().__init__()
+        self.exponent = exponent
+
+        # Precompute constants
+        p = torch.tensor(exponent + 1, dtype=torch.int32)
+        self.register_buffer("p", p)
+        self.register_buffer("a", -((p + 1) * (p + 2)) / 2)
+        self.register_buffer("b", p * (p + 2))
+        self.register_buffer("c", -p * (p + 1) / 2)
+
+    def forward(self, inputs: torch.Tensor) -> torch.Tensor:
+        # Compute powers efficiently
+        inputs_p_minus1 = torch.pow(inputs, self.p - 1)
+        inputs_p = inputs_p_minus1 * inputs  # inputs ** self.p
+        inputs_p_plus1 = inputs_p * inputs  # inputs ** (self.p + 1)
+
+        # Envelope function divided by r
+        env_val = (
+            (1.0 / inputs)
+            + self.a * inputs_p_minus1
+            + self.b * inputs_p
+            + self.c * inputs_p_plus1
+        )
+
+        # Apply cutoff
+        env_val = torch.where(inputs < 1.0, env_val, torch.zeros_like(env_val))
+
+        return env_val
+
+
+class BesselBasisLayer(nn.Module):
+    """
+    Bessel Basis Layer as used in DimeNet++.
+    """
+
+    def __init__(
+        self,
+        number_of_radial_bessel_functions: int,
+        radial_cutoff: float,
+        envelope_exponent: int = 5,
+    ):
+        super().__init__()
+        self.number_of_radial_bessel_functions = number_of_radial_bessel_functions
+        self.register_buffer(
+            "inv_cutoff", torch.tensor(1.0 / radial_cutoff, dtype=torch.float32)
+        )
+        self.envelope = Envelope(envelope_exponent)
+
+        # Initialize frequencies at canonical positions
+        frequencies = torch.pi * torch.arange(
+            1, number_of_radial_bessel_functions + 1, dtype=torch.float32
+        )
+        self.frequencies = nn.Parameter(frequencies)  # Trainable parameter
+
+    def forward(self, d_ij: torch.Tensor) -> torch.Tensor:
+        # d_ij: Pairwise distances between atoms. Shape: (nr_pairs, 1)
+
+        # Scale distances
+        d_scaled = d_ij * self.inv_cutoff  # Shape: (nr_pairs, 1)
+
+        # Apply envelope
+        d_cutoff = self.envelope(d_scaled)  # Shape: (nr_pairs, 1)
```

Comment:
@wiederm This matches what we saw (and was commented) in the implementation of the envelope function, distances are normalized with the cutoff distance. Please check my commentary in the envelope function.